### PR TITLE
[core] Port util::Thread to the actor model

### PR DESCRIPTION
**Pull request by tmpsantos**
Actor model version of util::Thread.

### CODE REVIEW

**Comment by tmpsantos** [comment]
My initial approach is to create a drop-in replacement for the `util::Thread` using the actor model as backend. After that we need a major redesign of the FileSource API to be something like subscribe/unsubscribe. @ivovandongen I need to see how you plan to use this inside the workers, but I think something like `FileSource::subscribe(Resource, ActorRef<FileSourceClient>)` sounds reasonable.

**Comment by tmpsantos** [comment]
@ivovandongen huge progress today, I wrote tons of unit tests, including one exchanging messages between the `ThreadPool` and the `ThreadedObject`. Only thing missing is to migrate a few parts of Android using `util::Thread`.

**Comment by tmpsantos** [comment]
@ivovandongen :eyes: ? This is the first step, now we need a better interface for the `FileSource`. But better in another PR otherwise this one would grow too much.

**Comment by tmpsantos** [comment]
:eyes: @jfirebaugh @kkaefer

File: src/mbgl/storage/file_source_request.hpp [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    void onCancel(std::function<void()>&& callback);
+    void setResponse(const Response& res);
+
+private:
+    void runResponseCallback();
+
+    std::shared_ptr<Mailbox> mailbox;
+
+    FileSource::Callback responseCallback = [](Response) {};
+    std::function<void()> cancelCallback = []{};
```
Comment (kkaefer) [inline src/mbgl/storage/file_source_request.hpp]: Having the cancel callback in here means that we're allocating more space for callbacks that are only used in one implementation.

File: platform/default/local_file_source.cpp [hunk 2]
```diff
@@ -59,26 +60,17 @@ LocalFileSource::LocalFileSource()
...
-    class LocalFileRequest : public AsyncRequest {
-    public:
-        LocalFileRequest(Resource resource_, FileSource::Callback callback_, ActorRef<LocalFileSource::Impl> fs_)
-            : mailbox(std::make_shared<Mailbox>(*util::RunLoop::Get()))
-            , fs(fs_) {
-            fs.invoke(&LocalFileSource::Impl::request, resource_.url, [callback_, ref = ActorRef<LocalFileRequest>(*this, mailbox)](Response res) mutable {
-                ref.invoke(&LocalFileRequest::runCallback, callback_, res);
-            });
-        }
+    auto mailbox = std::make_shared<Mailbox>(*util::RunLoop::Get());
```
Comment (kkaefer) [inline platform/default/local_file_source.cpp]: Can we share a Mailbox among all requests for a particular file source? `sizeof(Mailbox)` is 272 bytes on x64, and it seems fairly complex to construct.

File: src/mbgl/storage/file_source_request.hpp [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    ~FileSourceRequest() final;
+
+    void onResponse(FileSource::Callback&& callback);
+    void onCancel(std::function<void()>&& callback);
+    void setResponse(const Response& res);
+
+private:
+    void runResponseCallback();
+
+    std::shared_ptr<Mailbox> mailbox;
```
Comment (kkaefer) [inline src/mbgl/storage/file_source_request.hpp]: We're only storing this to make sure it continues to live as long as the request lives? If the Mailbox lived as long as the surrounding `*FileSource`, we could get rid of this pointer.

File: src/mbgl/storage/file_source_request.cpp [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+void FileSourceRequest::onCancel(std::function<void()>&& callback) {
+    cancelCallback = std::move(callback);
+}
+
+void FileSourceRequest::setResponse(const Response& response) {
+    // Copy, because calling the callback will
+    // sometimes self destroy this object.
+    auto callback = responseCallback;
+    callback(response);
```
Comment (tmpsantos) [inline src/mbgl/storage/file_source_request.cpp]: It can be called again in case the tile expires and gets re-requested by the `FileSource`. Or you get a response from the cache and later another one from the network.

File: src/mbgl/storage/file_source_request.hpp (hunk 1 above)
Comment (tmpsantos) [inline src/mbgl/storage/file_source_request.hpp]: Good point, gonna use `optional<>` here.

File: platform/default/default_file_source.cpp [hunk 5]
```diff
@@ -189,8 +189,8 @@ DefaultFileSource::DefaultFileSource(const std::string& cachePath,
                                      std::unique_ptr<FileSource>&& assetFileSource_,
                                      uint64_t maximumCacheSize)
         : assetFileSource(std::move(assetFileSource_))
-        , thread(std::make_unique<util::Thread<Impl>>(util::ThreadContext{"DefaultFileSource", util::ThreadPriority::Low},
```
Comment (tmpsantos) [inline platform/default/default_file_source.cpp]: All the `util::Thread` objects were setting priority to low, so I moved it to set to low on the `util::ThreadedObject` by default.

File: src/mbgl/storage/file_source_request.hpp (hunk 1 above)
Comment (kkaefer) [inline src/mbgl/storage/file_source_request.hpp]: You can use `nullptr` as "empty", then check `if (callback)`

File: platform/default/default_file_source.cpp (hunk 5 above)
Comment (tmpsantos) [inline platform/default/default_file_source.cpp]: I will better document this.

File: src/mbgl/storage/file_source_request.hpp (hunk 3 above)
Comment (tmpsantos) [inline src/mbgl/storage/file_source_request.hpp]: Let's do like you suggested. I have the `Mailbox` on the constructor because right now we assume the scheduler to be a `RunLoop` but that will have to change if we want to make requests from a worker in a `ThreadPool`. Better keep that to another PR.

File: src/mbgl/storage/file_source_request.hpp [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+namespace mbgl {
+
+class Mailbox;
+
+class FileSourceRequest : public AsyncRequest {
+public:
+    FileSourceRequest(std::shared_ptr<Mailbox> mailbox);
+    ~FileSourceRequest() final;
+
+    void onResponse(FileSource::Callback&& callback);
```
Comment (tmpsantos) [inline src/mbgl/storage/file_source_request.hpp]: Aesthetics to keep it consistent with `onCancel`, but lets change that.

File: platform/default/default_file_source.cpp [hunk 7]
```diff
@@ -237,22 +238,18 @@ void DefaultFileSource::setResourceTransform(std::function<std::string(Resource:
...
-        ~DefaultFileRequest() override {
-            thread.invoke(&DefaultFileSource::Impl::cancel, this);
-        }
+    request->onResponse(std::move(callback));
+    request->onCancel([fs = impl->actor(), req = request.get()] () mutable { fs.invoke(&Impl::cancel, req); });
 
-        util::Thread<DefaultFileSource::Impl>& thread;
-        std::unique_ptr<AsyncRequest> workRequest;
-    };
+    thread->invoke(&Impl::request, request.get(), resource, [ref =
```
Comment (tmpsantos) [inline platform/default/default_file_source.cpp]: Great idea, this is not only more Actor-ish but I could get rid of lambdas altogether on all file sources but the default.

**Comment by tmpsantos** [comment]
> Cool, I think this enables removing RunLoop::invokeWithCallback and WorkTask::makeWithCallback, which should make it net negative LOC.

Almost, I'm working on a follow-up PR refactoring `DefaultFileSource::setResourceTransform` that is access the `RunLoop` directly instead of an `ActorRef`. After that we can remove these methods.

Thanks @kkaefer @jfirebaugh @ivovandongen for reviewing. Looks much better now.

File: platform/default/local_file_source.cpp (hunk 2 above)
Comment (tmpsantos) [inline platform/default/local_file_source.cpp]: I don't think it is possible. We share the `Scheduler`, but sharing a `Mailbox` would imply that the object would still be scheduled after its destruction. But thanks for pointing this out, because I just realized we are not closing the `Mailbox` on the `FileSourceRequest` destructor.
